File: generate/reference/duality-of-compilation/ocaml/erase_unused.ml
let rec replicate v n a =
  if n = 0 then a
  else replicate v (n - 1) (v :: a)

let rec useless i n _ =
  if i < n then useless (i + 1) n (replicate 0 i [])
  else i

let main n =
  useless 0 n []

let () =
  match Sys.argv with
  | [| _ |] -> print_endline (string_of_int (main 10))
  | [| _; s |] -> (
      match int_of_string_opt s with
      | Some n -> print_endline (string_of_int (main n))
      | None -> print_endline "argument is not a number")
  | _ -> print_endline "arguments are too many"

File: generate/reference/duality-of-compilation/rust/erase_unused.rs
use std::env;
use std::rc::Rc;

enum List {
    Nil,
    Cons(isize, Rc<List>),
}

fn replicate(value: isize, n: isize) -> List {
    let mut list = List::Nil;
    for _ in 0..n {
        list = List::Cons(value, Rc::new(list));
    }
    list
}

fn useless(n: isize) -> isize {
    let mut i = 0;
    for j in 0..n {
        replicate(0, j);
        i += 1;
    }
    i
}

fn run(n: isize) -> isize {
    useless(n)
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let arg = match args.get(1) {
        None => 10,
        Some(val) => match val.parse::<isize>() {
            Ok(val) => val,
            Err(_) => {
                println!("argument is not a number");
                return;
            }
        },
    };

    println!("{}", run(arg));
}
